### ISSUE TITLE
v0.30.0, JPEG input can occasionally trigger shrink-on-load when it shouldn't

### ISSUE THREAD

**Issue by adityapatadia** [lovell/sharp#3066]
Here are image samples which are processed with 0.29.3 vs 0.30.0. The output with latest version is more blurry. Also when resized to 500 pixel width, 0.29.3 outputs 334 pixel height while 0.30 outputs 333 pixel height.

I also noticed that byte size of image given by 0.30 is smaller.

Original:
![original](https://user-images.githubusercontent.com/1086617/152109447-4bc0fb67-1c3a-434a-877c-cb8c078d52e8.jpeg)

0.29.3
![sharp-0 29 0](https://user-images.githubusercontent.com/1086617/152109460-427ce0b1-a6db-4b46-a018-32fadfdb1bb7.jpeg)

0.30.0
![sharp-0 30 0](https://user-images.githubusercontent.com/1086617/152109461-44e9cd35-2821-4c90-af44-06377b5f2e61.jpeg)
.

**Comment by adityapatadia** [comment]
I would also like to mention this happens regardless of output format. I think some issue is there in image resize functionality.

**Comment by lovell** [comment]
It looks like this occurs only for JPEG input where an (even) output dimension is exactly half the input dimension.

In v0.30.0 this is triggering shrink-on-load, which for the sample image results in the (known) rounding error of 500x333 as the target dimensions are 500x334. This means there's a slight upsample (vscale=1.001502) to attempt to increase the height, hence the "blur".

Using `.resize({ width: 500, fastShrinkOnLoad: false })` with the sample image works as v0.29.3.

@kleisauke I guess we might want to make the following change to avoid the use of shrink on load with these exact half/quarter reductions. Or alternatively we could skip the later upsample and live with the shrink-on-load rounding error?

```diff
--- a/src/pipeline.cc
+++ b/src/pipeline.cc
@@ -186,11 +186,11 @@ class PipelineWorker : public Napi::AsyncWorker {
           // Leave at least a factor of two for the final resize step, when fastShrinkOnLoad: false
           // for more consistent results and avoid occasional small image shifting
           int factor = baton->fastShrinkOnLoad ? 1 : 2;
-          if (shrink >= 8 * factor) {
+          if (shrink > 8 * factor) {
             jpegShrinkOnLoad = 8;
-          } else if (shrink >= 4 * factor) {
+          } else if (shrink > 4 * factor) {
             jpegShrinkOnLoad = 4;
-          } else if (shrink >= 2 * factor) {
+          } else if (shrink > 2 * factor) {
             jpegShrinkOnLoad = 2;
           }
```

**Comment by kleisauke** [comment]
Ah, indeed, this is a slight oversight on my part. Probably commit https://github.com/lovell/sharp/commit/e398b471e1e69c789bfda76e7a9f3e87ecfaf3b7 is also relevent here. How about doing this instead?:
```diff
--- a/src/pipeline.cc
+++ b/src/pipeline.cc
@@ -193,6 +193,10 @@ class PipelineWorker : public Napi::AsyncWorker {
           } else if (shrink >= 2 * factor) {
             jpegShrinkOnLoad = 2;
           }
+          // Help ensure a final kernel-based reduction to prevent shrink aliasing
+          if (jpegShrinkOnLoad > 1 && shrink == jpegShrinkOnLoad) {
+            jpegShrinkOnLoad /= 2;
+          }
         } else if (inputImageType == sharp::ImageType::WEBP ||
                    inputImageType == sharp::ImageType::SVG ||
                    inputImageType == sharp::ImageType::PDF) {
```

As I think swapping `>=` with `>` would not match libvips' behaviour in case `fastShrinkOnLoad` option is set to false.
https://github.com/libvips/libvips/blob/d9c00eb5ca577018b40533f041230754484ace7e/libvips/resample/thumbnail.c#L487-L501

FWIW, here's a handy way to check if it would match `vipsthumbnail`'s behavior.
```bash
$ vips copy x.jpg x-stripped.jpg[strip]
$ node -e "require('sharp')('x-stripped.jpg').resize({ width: 500, fastShrinkOnLoad: false }).toFile('sharp.jpg')"
$ vipsthumbnail x-stripped.jpg -s 500x -o vips.jpg[strip,Q=80,subsample_mode=on,optimize_coding]
$ sha256sum sharp.jpg vips.jpg
7c765ad7fec06e9e32c9a85ec0da6f705be212c75791e96b32ed1705bb740dba  sharp.jpg
7c765ad7fec06e9e32c9a85ec0da6f705be212c75791e96b32ed1705bb740dba  vips.jpg
```

**Comment by lovell** [comment]
Commit https://github.com/lovell/sharp/commit/2b019513068fb596c9ff1760266306b239e6aed9 adds these dimensions as a test case, plus Kleis' fix.

**Comment by kleisauke** [comment]
After some further investigation, it looks like this workaround:
https://github.com/lovell/sharp/blob/424660278dfbb1f6cfb146aec1d1039714a05b1c/src/pipeline.cc#L284-L290

Can also be resolved by casting the shrink factor to an integer before the comparison. See for example commit https://github.com/kleisauke/sharp/commit/62d95b1b76e995aa6650c7a9ba55cfdf8bece9de. This would also ensure that `fastShrinkOnLoad` would behave similarly (i.e. no shifts or moiré patterns) when that option is turned off, without sacrificing much performance (since only the edge cases would lower the shrink-on-load factor). For example, I noticed that these 2 tests:
https://github.com/lovell/sharp/blob/424660278dfbb1f6cfb146aec1d1039714a05b1c/test/unit/resize.js#L640
https://github.com/lovell/sharp/blob/424660278dfbb1f6cfb146aec1d1039714a05b1c/test/unit/resize.js#L652
Are now producing the identical image with that commit.

**Comment by lovell** [comment]
Oh that's much neater, thanks Kleis, a PR would be most welcome.

**Comment by lovell** [comment]
@kleisauke I landed your patch via commit https://github.com/lovell/sharp/commit/7faacd91b0d7f59b71d8c3ecfb3a0c9936c4999d

**Comment by lovell** [comment]
v0.30.1 now available, thanks for the fix Kleis.

**Comment by StephanHoyer** [comment]
I still see this issue when I scale a 1600x1000px to width 10px which should result in 10 * 1.6 = 6.25 ≈ 6

source:
![test](https://user-images.githubusercontent.com/54701/153355340-eadc3539-b654-4e86-8d6c-ae8d12e6da62.jpg)

0.29.0: height 6px
![ae3c87692818dd757f725717f60741bd1fae0e3a764fa62ad628e7fab89e5053](https://user-images.githubusercontent.com/54701/153356421-d131591d-6cbf-48be-a252-a843870b3aa8.jpg)


0.30.1: height 7px
![ae3c87692818dd757f725717f60741bd1fae0e3a764fa62ad628e7fab89e5053](https://user-images.githubusercontent.com/54701/153356161-373d2833-f98a-4de2-8ac8-945b2b0bd59f.jpg)

**Comment by kleisauke** [comment]
@StephanHoyer thanks for the test case. I was able to reproduce that with the `vipsthumbnail` utility:
```bash
$ vips black x.jpg 1600 1000
$ vipsthumbnail x.jpg -s 10x -o vips.jpg
$ vipsheader vips.jpg
vips.jpg: 10x7 uchar, 1 band, b-w, jpegload
``` 

PR https://github.com/libvips/libvips/pull/1769 should resolve that.

**Comment by StephanHoyer** [comment]
FYI: We still see this issue.

**Comment by kleisauke** [comment]
@StephanHoyer This was fixed in libvips v8.13.0, which will ship with the forthcoming sharp v0.31.0 - see: https://www.libvips.org/2022/05/28/What's-new-in-8.13.html#image-resize-quality-improvements
